src/datasets/vision/imagenet_reader/preprocess.jl

```diff
@@ -0,0 +1,31 @@
+# Image preprocessing for ImageNet models.
+# Code adapted from Metalhead 0.5.3's utils.jl
+
+# Take rectangle of pixels of shape `outsize` at the center of image `im`
+adjust(i::Integer) = ifelse(iszero(i % 2), 1, 0)
+function center_crop_view(im::AbstractMatrix, outsize=IMGSIZE)
+    h2, w2 = div.(outsize, 2) # half height, half width of view
+    h_adjust, w_adjust = adjust.(outsize)
+    return @view im[
+        ((div(end, 2) - h2):(div(end, 2) + h2 - h_adjust)) .+ 1,
+        ((div(end, 2) - w2):(div(end, 2) + w2 - w_adjust)) .+ 1,
+    ]
+end
+
+# Coefficients taken from PyTorch's ImageNet normalization code
+const PYTORCH_MEAN = [0.485f0, 0.456f0, 0.406f0]
+const PYTORCH_STD = [0.229f0, 0.224f0, 0.225f0]
+
+function preprocess(Tx::Type, im::AbstractMatrix{<:AbstractRGB})
+    im = center_crop_view(im)
+    im = (channelview(im) .- PYTORCH_MEAN) ./ PYTORCH_STD
```

Comment:
I took another look at the PyTorch example. When creating a dataset, they allow a `transform` function to be passed, so the normalisation is indeed not part of the dataset:

https://github.com/pytorch/examples/blob/91ccd7a21be6fa687000beef82fc1e5d7d64e4bd/imagenet/main.py#L211-L230
```python
normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406],
                                 std=[0.229, 0.224, 0.225])

train_dataset = datasets.ImageFolder(
    traindir,
    transforms.Compose([
        transforms.RandomResizedCrop(224),
        transforms.RandomHorizontalFlip(),
        transforms.ToTensor(),
        normalize,
    ]))

val_dataset = datasets.ImageFolder(
    valdir,
    transforms.Compose([
        transforms.Resize(256),
        transforms.CenterCrop(224),
        transforms.ToTensor(),
        normalize,
    ]))
```

More interestingly, all of the resizing and cropping also isn't "hardcoded" into the data loader.

Reply:
Yes we should remove them

Reply:
I do think having a standarized way to preprocess should be distributed with the DataSet, at least in the form of a Readme (e.g. by posting the code above), so that it is easy for people to get started.

Reply:
[09feb3d](https://github.com/JuliaML/MLDatasets.jl/pull/146/commits/09feb3d4a621df3fca6e38f3fb9119b6bcb9ac4b) adds this feature to the constructor:
```Julia
function ImageNet(
    Tx::Type,
    split::Symbol;
    preprocess=ImageNetReader.default_preprocess,
    inverse_preprocess=ImageNetReader.default_inverse_preprocess,
    # ...
)